src/sentry/rules/history/preview.py

```diff
@@ -198,34 +199,66 @@ def get_fired_groups(
 
 
 def get_top_groups(
-    project: Project, start: datetime, end: datetime, condition_activity: GroupActivityMap
+    project: Project,
+    start: datetime,
+    end: datetime,
+    condition_activity: GroupActivityMap,
+    dataset_map: Dict[int, Dataset],
 ) -> GroupActivityMap:
     """
     Filters the activity to contain only groups that have the most events in the past 2 weeks.
 
     Since frequency conditions require one snuba query per groups, we need to limit the number groups we process.
     """
-    if not condition_activity:
-        return condition_activity
-    # TODO: Also check other datasets for top groups
-    groups = raw_query(
-        dataset=Dataset.Events,
-        start=start,
-        end=end,
-        filter_keys={"project_id": [project.id]},
-        conditions=[("group_id", "IN", list(condition_activity.keys()))],
-        aggregations=[("count", "group_id", "groupCount")],
-        groupby=["group_id"],
-        order_by="-groupCount",
-        seleted_columns=["group_id"],
-        limit=FREQUENCY_CONDITION_GROUP_LIMIT,
-    ).get("data", [])
-
-    top_groups = {group["group_id"] for group in groups}
-    top_activity = {
+    datasets = {dataset_map.get(group) for group in condition_activity.keys()}
+
+    # queries each dataset for top x groups and then gets top x overall
+    query_params = []
+    for dataset in datasets:
+        if dataset is None:
+            continue
+        group_ids = list(condition_activity.keys())
+        kwargs = {
+            "dataset": dataset,
+            "start": start,
+            "end": end,
+            "filter_keys": {"project_id": [project.id]},
+            "conditions": [("group_id", "IN", group_ids)],
+            "aggregations": [("count", "group_id", "groupCount")],
+            "groupby": ["group_id"],
+            "order_by": "-groupCount",
+            "seleted_columns": ["group_id"],
+            "limit": FREQUENCY_CONDITION_GROUP_LIMIT,
+        }
+        if dataset == Dataset.Transactions:
+            kwargs["conditions"] = [[["hasAny", ["group_ids", ["array", group_ids]]], "=", 1]]
+            kwargs["aggregations"].append(("arrayJoin", ["group_ids"], "group_id"))
+        query_params.append(SnubaQueryParams(**kwargs))
```

Comment:
So does using `filter_keys` instead of the IN `condition` filter the rows before arrayjoining?

Reply:
Yeah, I believe `filter_keys` is really an alias for conditions except `conditions` is more general. Not sure why it was done this way - I suspect for usability of the function. 

With querying for transactions, you wanna avoid any conditions on `group_id` since that will force the arrayJoin on `group_ids` -> `group_id`. That's why in my query for transactions I completely removed the filter_keys predicate and the `[("group_id", "IN", group_ids)]` condition.

Hope that makes sense.

Reply:
If the dataset is transactions we're replacing the `IN` condition with the `hasAny` which I thought would filter out the unneeded arrayJoins. It looks like the difference here is that for the non-transaction case you're replacing the `IN` condition with the filter_key predicate?